File: gitlfs-client/src/main/java/ru/bozaro/gitlfs/client/AuthHelper.java
package ru.bozaro.gitlfs.client;

import ru.bozaro.gitlfs.client.auth.AuthProvider;
import ru.bozaro.gitlfs.client.auth.BasicAuthProvider;
import ru.bozaro.gitlfs.client.auth.ExternalAuthProvider;

import javax.annotation.Nonnull;
import java.net.MalformedURLException;
import java.net.URI;
import java.net.URISyntaxException;

/**
 * Utility class.
 *
 * @author Artem V. Navrotskiy
 */
public final class AuthHelper {
  private AuthHelper() {
  }

  /**
   * Create AuthProvider by gitURL.
   * <p>
   * Supported URL formats:
   * <p>
   * * https://user:passw0rd@github.com/foo/bar.git
   * * http://user:passw0rd@github.com/foo/bar.git
   * * git://user:passw0rd@github.com/foo/bar.git
   * * ssh://git@github.com/foo/bar.git
   * * git@github.com:foo/bar.git
   * <p>
   * Detail Git URL format: https://git-scm.com/book/ch4-1.html
   *
   * @param gitURL URL to repository.
   * @return Created auth provider.
   */
  @Nonnull
  public static AuthProvider create(@Nonnull String gitURL) throws MalformedURLException {
    if (gitURL.contains("://")) {
      final URI uri = URI.create(gitURL);
      final String path = uri.getPath();
      switch (uri.getScheme()) {
        case "https":
        case "http":
        case "git":
          return new BasicAuthProvider(join(uri, "info/lfs"));
        case "ssh":
          return new ExternalAuthProvider(uri.getAuthority(), path.startsWith("/") ? path.substring(1) : path);
        default:
          throw new MalformedURLException("Can't find authenticator for scheme: " + uri.getScheme());
      }
    }
    return new ExternalAuthProvider(gitURL);
  }

  @Nonnull
  public static URI join(@Nonnull URI href, @Nonnull String... path) {
    try {
      URI uri = new URI(href.getScheme(), href.getAuthority(), href.getPath() + (href.getPath().endsWith("/") ? "" : "/"), null, null);
      for (String fragment : path)
        uri = uri.resolve(fragment);
      return uri;
    } catch (URISyntaxException e) {
      throw new IllegalStateException(e);
    }
  }
}

File: gitlfs-client/src/main/kotlin/ru/bozaro/gitlfs/client/AuthHelper.kt
package ru.bozaro.gitlfs.client

import ru.bozaro.gitlfs.client.auth.AuthProvider
import ru.bozaro.gitlfs.client.auth.BasicAuthProvider
import ru.bozaro.gitlfs.client.auth.ExternalAuthProvider
import java.net.MalformedURLException
import java.net.URI
import java.net.URISyntaxException

/**
 * Utility class.
 *
 * @author Artem V. Navrotskiy
 */
object AuthHelper {
    /**
     * Create AuthProvider by gitURL.
     *
     *
     * Supported URL formats:
     *
     *
     * * https://user:passw0rd@github.com/foo/bar.git
     * * http://user:passw0rd@github.com/foo/bar.git
     * * git://user:passw0rd@github.com/foo/bar.git
     * * ssh://git@github.com/foo/bar.git
     * * git@github.com:foo/bar.git
     *
     *
     * Detail Git URL format: https://git-scm.com/book/ch4-1.html
     *
     * @param gitURL URL to repository.
     * @return Created auth provider.
     */
    @kotlin.jvm.JvmStatic
    @Throws(MalformedURLException::class)
    fun create(gitURL: String): AuthProvider {
        if (gitURL.contains("://")) {
            val uri = URI.create(gitURL)
            val path = uri.path
            return when (uri.scheme) {
                "https", "http", "git" -> BasicAuthProvider(join(uri, "info/lfs"))
                "ssh" -> ExternalAuthProvider(uri.authority, if (path.startsWith("/")) path.substring(1) else path)
                else -> throw MalformedURLException("Can't find authenticator for scheme: " + uri.scheme)
            }
        }
        return ExternalAuthProvider(gitURL)
    }

    fun join(href: URI, vararg path: String): URI {
        return try {
            var uri = URI(href.scheme, href.authority, href.path + if (href.path.endsWith("/")) "" else "/", null, null)
            for (fragment in path) uri = uri.resolve(fragment)
            uri
        } catch (e: URISyntaxException) {
            throw IllegalStateException(e)
        }
    }
}
